src/simulations/collect.jl

```diff
@@ -0,0 +1,356 @@
+  else
+    df = run!(model, agent_step!, model_step!, n;
+    when=when, agent_properties=agent_properties,
+    model_properties=model_properties, aggregation_dict=aggregation_dict)
+    return df
+  end
+end
+
+###################################################
+# core data collection functions per step
+###################################################
+"""
+    collect_agent_data(model::ABM, properties::Vector, step = 0) → df
+
+Collect agent properties into a dataframe. `properties` can have symbols (fields) or
+functions that take an agent as input and output a value. `step` is given only
+so that the function adds the correct number in the `step` column.
+"""
+function collect_agent_data(model::ABM, properties::AbstractArray, step::Int = 0)
+  dd = DataFrame()
+  dd[!, :id] = collect(keys(model.agents))
+  for fn in properties
+    dd[!, Symbol(fn)] = get_data.(values(model.agents), fn)
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
+collect_agent_data(model::ABM, properties::Nothing, step::Int=0) = DataFrame()
+
+"""
+    collect_model_data(model::ABM, properties::Vector, step = 0) → df
+
+Same as [`collect_agent_data`](@ref) but for model data instead.
+"""
+function collect_model_data(model::ABM, properties::AbstractArray, step::Int = 0)
+  dd = DataFrame()
+  for fn in properties
+    r =  get_data(model, fn)
+    dd[!, Symbol(fn)] = [r]
+  end
+  dd[!, :step] = fill(step, size(dd, 1))
+  return dd
+end
+
+collect_model_data(model::ABM, properties::Nothing, step::Int=0) = DataFrame()
+
+"""
+    aggregate_data(df::AbstractDataFrame, aggregation_dict::Dict)
+
+Aggregate `df` columns  with some function(s) specified in `aggregation_dict`.
+Each key in `aggregation_dict` is a column name (Symbol), and each value is
+an array of function to aggregate that column.
+
+Aggregation occurs per step.
+"""
+function aggregate_data(df::AbstractDataFrame, aggregation_dict::Dict)
+  all_keys = collect(keys(aggregation_dict))
+  dfnames = names(df)
+  available_keys = [k for k in all_keys if in(k, dfnames)]
+  length(available_keys) == 0 && return
+  v1 = aggregation_dict[available_keys[1]]
+  final_df = by(df, :step,  available_keys[1] => v1[1])
+  for v2 in v1[2:end]
+    dd = by(df, :step,  available_keys[1] => v2)
+    final_df = join(final_df, dd, on=:step)
+  end
+  for k in available_keys[2:end]
+    v = aggregation_dict[k]
+    for v2 in v
+      dd = by(df, :step,  k => v2)
+      final_df = join(final_df, dd, on=:step)
+    end
+  end
+
+  # rename columns
+  colnames = Array{Symbol}(undef, size(final_df, 2))
+  colnames[1] = :step
+  counter = 2
+  for k in available_keys
+    v = aggregation_dict[k]
+    for vv in v
+      colnames[counter] = Symbol(join([vv,"(", string(k), ")"], ""))
+      counter += 1
+    end
+  end
+  rename!(final_df, colnames)
+
+  return final_df
+end
+
+"used in _run!"
+function collect_agent_data!(df::DataFrame, model::ABM, agent_properties,  aggregation_dict, step::Int)
```

Comment:
This is not at in-place function, so shouldn't have an `!` in its name.

Reply:
this would confilct with the version that initializes the data container.

Reply:
It is updating `df`.

Reply:
Not currently, no. It returns a new dataframe, but the input `df` remains as it was before the call.

Anyhow, this will change with the new implementations we've discussed in slack.

Reply:
`df = vcat(df, dfa)` this updates the `df`.

Reply:
But it doesn't. `vcat` is not an in-place function and the `df` result is now locally scoped. To update in place there would be no need to return a dataframe and we'd expect the method to alter the input.

```julia
julia> df_agent = DataFrame();
julia> df = Agents.collect_agent_data!(df_agent, model, [:weight], nothing, 1);
julia> df_agent
0×0 DataFrame
julia> df
3×3 DataFrame
│ Row │ id    │ weight   │ step  │
│     │ Int64 │ Float64  │ Int64 │
├─────┼───────┼──────────┼───────┤
│ 1   │ 2     │ 0.300242 │ 1     │
│ 2   │ 3     │ 0.729594 │ 1     │
│ 3   │ 1     │ 0.145448 │ 1     │
```

Reply:
Right. I hadn't tested the method and I overlooked the local scope.